### PR TITLE
Refactor visibility to ElasticSearch 

### PR DESCRIPTION
**Pull request by vancexu**
https://github.com/uber/cadence/issues/1718

### CODE REVIEW

File: common/persistence/visibilityStore.go [hunk 1]
```diff
@@ -192,3 +220,14 @@ func (v *visibilityManagerImpl) convertVisibilityWorkflowExecutionInfo(execution
 
 	return convertedExecution
 }
+
+func (v *visibilityManagerImpl) getMemo(visibilityMemo *shared.Memo, domainID, wID, rID string) *DataBlob {
```
Comment (yycptt) [inline common/persistence/visibilityStore.go]: Maybe change the name to serializeMemo or getSerializedMemo?

File: service/history/timerQueueProcessorBase.go [hunk 2]
```diff
@@ -711,31 +706,10 @@ func (t *timerQueueProcessorBase) deleteWorkflowHistory(task *persistence.TimerT
...
-func getVisibilityMessageForDeletion(domainID, workflowID, runID string, docVersion int64) *indexer.Message {
-	msgType := indexer.MessageTypeDelete
-	msg := &indexer.Message{
-		MessageType: &msgType,
-		DomainID:    common.StringPtr(domainID),
-		WorkflowID:  common.StringPtr(workflowID),
-		RunID:       common.StringPtr(runID),
-		Version:     common.Int64Ptr(docVersion),
+	op := func() error {
+		return t.historyService.DeleteExecutionFromVisibility(task)
```
Comment (yycptt) [inline service/history/timerQueueProcessorBase.go]: If deletion from es is successful but deletion from db failed, will the deletion from es fail in the next attempt?

File: common/persistence/visibilityStore.go [hunk 3]
```diff
@@ -192,3 +220,14 @@ func (v *visibilityManagerImpl) convertVisibilityWorkflowExecutionInfo(execution
...
+	memo, err := v.serializer.SerializeVisibilityMemo(visibilityMemo, VisibilityEncoding)
+	if err != nil {
+		logging.LogErrorEncodingVisibility(v.logger, err, domainID, wID, rID)
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/visibilityStore.go]: You probably need to rebase the latest master. logging is changed to new interface and those helpers and bark are gone.

[the remark this answers is not in the document]
Comment (vancexu) [reply common/persistence/visibilityStore.go]: > This logging requires lots of parameters which is not related to memo serialization. How about doing the logging outside this function to keep it simple?

Doing outside caused duplicate codes in every caller, so prefer this way. (An alternative/comprised is create a struct from domain, wid, rid, but I see little motivation for that change.)

 

> You probably need to rebase the latest master. logging is changed to new interface and those helpers and bark are gone.
Rebased. Tons of conflicts.

File: common/persistence/elasticsearch/esVisibilityManager.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/uber/cadence/common/metrics"
+	p "github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/common/service/config"
+	"github.com/uber/cadence/common/tokenbucket"
+)
+
+// NewESVisibilityManager create a visibility manager for ElasticSearch
+// In history, it only needs kafka producer for writing data;
+// In frontend, it only needs ES client and related config for reading data
+func NewESVisibilityManager(indexName string, esClient es.Client, config *config.VisibilityConfig,
```
[the remark this answers is not in the document]
Comment (vancexu) [reply common/persistence/elasticsearch/esVisibilityManager.go]: agree, updated

File: common/persistence/visibilityStore.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (vancexu) [reply common/persistence/visibilityStore.go]: ok

File: service/history/timerQueueProcessorBase.go (hunk 2 above)
Comment (vancexu) [reply service/history/timerQueueProcessorBase.go]: No it won't. It just publish a duplicate deletion event to kafka, and indexer to ES will handle such duplicates.

File: service/history/service.go [hunk 5]
```diff
@@ -305,7 +317,7 @@ func (s *Service) Start() {
 		log.Fatalf("Creating historyV2 manager persistence failed: %v", err)
 	}
 
-	handler := NewHandler(base, s.config, shardMgr, metadata, visibility, history, historyV2, pFactory, params.PublicClient)
+	handler := NewHandler(base, s.config, shardMgr, metadata, visibility, esVisibility, history, historyV2, pFactory, params.PublicClient)
```
[the remark this answers is not in the document]
Comment (vancexu) [reply service/history/service.go]: Good point, updated

**Pull request merged**